### ISSUE TITLE
Change search result ranking

### ISSUE THREAD

**Issue by kineticsquid** [EclipseFdn/open-vsx.org#3194]
I recently received a complaint about an extension that was misleading. It was a republication of a popular extension, sight changes to `package.json`, but otheriwse no changes to name, displayname or any of the links. And because it was new, it was showing up first in the search results, ahead of the much more popular extension. I'm recommending we change our search ranking to more heavily, or completely, weight on download count. 

@filiptronicek @spoenemann @amvanbaren Thoughts?

**Comment by filiptronicek** [comment]
@amvanbaren did some changes to sorting last year in https://github.com/eclipse/openvsx/pull/672, but IIRC that was just for sorting by "Rating". I'm not sure what we sort by now by default, but download count should definitely be one of the main factors we take a look at (besides probably string distance).

**Comment by amvanbaren** [comment]
By default it sorts by relevance, which is a combination of multiple weighting factors: https://github.com/eclipse/openvsx/blob/072bf3e7fa38eef217ebb30d9748f5556440c665/server/src/main/java/org/eclipse/openvsx/search/RelevanceService.java#L39-L48

This is how the relevance rating is calculated: https://github.com/eclipse/openvsx/blob/072bf3e7fa38eef217ebb30d9748f5556440c665/server/src/main/java/org/eclipse/openvsx/search/RelevanceService.java#L73-L109

The `timestamp` factor is the leading relevance factor for the open-vsx.org deployment: https://github.com/EclipseFdn/open-vsx.org/blob/43c654cf6e567bb720749d4323996bc23598e81b/configuration/application.yml#L177-L181

**Comment by kineticsquid** [comment]
IMO downloads should be most heavily weighted. Not sure of the values though.

**Comment by filiptronicek** [comment]
I'd propose that we switch the open-vsx.org deployment to the following:
```
 rating: 0.5
 downloads: 2.0
 timestamp: 1.2
```

This is as arbitrary as it gets, but it feels like we should make ratings and downloads the primary metrics and only have the timestamp secondary, since besides indicating that an extension is more "up-to-date", it does not signal much.

**Comment by kineticsquid** [comment]
I was going to wait until it was deployed.

**Comment by kineticsquid** [comment]
OK, https://github.com/EclipseFdn/open-vsx.org/pull/3206 was deployed with the ranking changes. I honestly don't recall the specific search result that caused us to make this change. That said, I tried searching on`java` with the following results:

<img width="679" alt="Image" src="https://github.com/user-attachments/assets/2a40e40a-c77d-46f0-ad5f-c610296ebaf0" />

The first result is as expected. However, `KylinIdeTeam/java` has fewer downloads and was published earlier than `vscode/java`. I would have expected that order switched.

**Comment by amvanbaren** [comment]
Yes, same for the `Spring Initializr` and `Maven for Java` extensions. I'll add some logging to debug the relevance calculation.

**Comment by kineticsquid** [comment]
@amvanbaren is https://github.com/eclipse/openvsx/issues/1269 to fix this? I'm seeing the same ordering of the Java extensions above.

**Comment by amvanbaren** [comment]
@kineticsquid No, it was to fix sorting by downloads, date or rating.

**Comment by kineticsquid** [comment]
@amvanbaren Could you net out the effects of the changes in the PR?

We'll need to hear from @chrisguindon and @tfroment.

**Comment by amvanbaren** [comment]
@kineticsquid The main difference is that it multiplies the ElasticSearch `score` with the `relevance` instead of using the `relevance` as a tie breaker.

**Comment by chrisguindon** [comment]
+ @netomi - I think @kineticsquid was trying to cc you on the issue but added @tfroment instead!

**Comment by solomatov** [comment]
I reported a couple of issues which might be relevant here:
- https://github.com/eclipse/openvsx/issues/1442
- https://github.com/eclipse/openvsx/issues/1443

**Comment by solomatov** [comment]
It seems that many issues might be related to fuzzyness, i.e. see this bug: https://github.com/EclipseFdn/open-vsx.org/issues/5260 May be you could experiment with turning on/off this setting?

**Comment by solomatov** [comment]
@kineticsquid
>IMO downloads should be most heavily weighted. Not sure of the values though.

If there's a fuzzy search which brings a lot of irrelevant results downloads will make the situation worse. See for example, the query C#: https://open-vsx.org/?search=C%23&sortBy=relevance&sortOrder=desc

It contains a lot of relevant stuff due to mix up of C++ and C#, since they seem to fuzzy match.

**Comment by kineticsquid** [comment]
@chrisguindon Thanks, yes, I was trying to cc @netomi

**Comment by netomi** [comment]
there are two things that factor into this imho:

 - search relevance calculation (should be improved by https://github.com/eclipse/openvsx/pull/1385)
 - fuzzy search by default (should be improved by https://github.com/eclipse/openvsx/pull/1579)

currently we do text search with fuzzyness by default, this means that exact and fuzzy matches receive the same boost and thus you get some really weird results.

The referenced PR does address this by splitting off exact and fuzzy matches such that fuzzy matches do not receive a boost anymore and we do not fuzzy match tags as this does not make sense imho.

**Comment by netomi** [comment]
a change to the way the relevance ranking is performed is deployed to staging.open-vsx.org

it uses the formula from imdb and results in much better relevance ranking as currently, tested with the current production database.
